NPE in `merge-fields*` when updating doc with schema change and multiple docs in same tx

(Initially found and minimized by #4719)

Running a transaction that updates a document’s schema and includes multiple docs for the same ID causes a NullPointerException in Arrow schema merging:

```
(t/deftest minimzed-sample-npe
  (with-open [node (xtn/start-node {:databases {:xtdb {:log [:in-memory {:instant-src (tu/->mock-clock)}]}}
                                                       :compactor {:threads 0}})]
    (xt/execute-tx node [[:put-docs :docs {:h {:b #{}} :xt/id 1}]])
    (xt/execute-tx node [[:put-docs :docs {:h {}, :xt/id 1} {:xt/id 1}]])
    (t/is (= [{:xt/id 1}] (xt/q node "SELECT * FROM docs ORDER BY _id")))))
```

Leads to the following error on queries against docs:
```
; xtdb.error.Fault: Cannot invoke "org.apache.arrow.vector.types.pojo.Field.getType()" because "field" is null
; 	at xtdb.error$fault.invokeStatic(error.clj:90)
; 	at xtdb.error$fault.invoke(error.clj:82)
; ...
; 	at java.base/java.lang.Thread.run(Thread.java:1583)
; Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.arrow.vector.types.pojo.Field.getType()" because "field" is null
; 	at xtdb.types$merge_fields_STAR_$merge_field_STAR___19493.invoke(types.clj:352)
; 	at xtdb.types$merge_fields_STAR_$merge_field_STAR___19493$fn__19495.invoke(types.clj:362)
; 	at clojure.core$update.invokeStatic(core.clj:6259)
; 	at clojure.core$update.invoke(core.clj:6251)
; 	at xtdb.types$merge_fields_STAR_$merge_field_STAR___19493.invoke(types.clj:357)
; 	at clojure.core$fnil$fn__6990.invoke(core.clj:6646)
; 	at clojure.core$update.invokeStatic(core.clj:6261)
; 	at clojure.core$update.invoke(core.clj:6251)
; 	at xtdb.types$merge_fields_STAR_$merge_field_STAR___19493$fn__19495$fn__19499.invoke(types.clj:368)
; 	at clojure.core.protocols$iterator_reduce_BANG_.invokeStatic(protocols.clj:42)
...
```

--- a/core/src/main/clojure/xtdb/types.clj
+++ b/core/src/main/clojure/xtdb/types.clj
@@ -358,9 +358,15 @@
                             (fn [{:keys [nullable?] :as type-opts}]
                               (into {:nullable? (or nullable? (.isNullable field))}
                                     (condp = (class arrow-type)
-                                      ArrowType$List {:el (merge-field* (:el type-opts) (first (.getChildren field)))}
-                                      SetType {:el (merge-field* (:el type-opts) (first (.getChildren field)))}
-                                      ArrowType$FixedSizeList {:el (merge-field* (:el type-opts) (first (.getChildren field)))}
+                                      ArrowType$List (let [child-field (first (.getChildren field))]
+                                                       {:el (cond-> (:el type-opts)
+                                                              child-field (merge-field* child-field))})
+                                      SetType (let [child-field (first (.getChildren field))]
+                                                {:el (cond-> (:el type-opts)
+                                                       child-field (merge-field* child-field))})
+                                      ArrowType$FixedSizeList (let [child-field (first (.getChildren field))]
+                                                                {:el (cond-> (:el type-opts)
+                                                                       child-field (merge-field* child-field))})
                                       ArrowType$Struct {:fields (let [default-field-mapping (if type-opts {#xt.arrow/type :null {:nullable? true}} nil)
                                                                       children (.getChildren field)]
                                                                   (as-> (:fields type-opts) fields-acc

--- a/src/test/clojure/xtdb/node_test.clj
+++ b/src/test/clojure/xtdb/node_test.clj
@@ -1,6 +1,6 @@
 (ns xtdb.node-test
   (:require [clojure.java.io :as io]
-            [clojure.test :as t :refer [deftest]] 
+            [clojure.test :as t :refer [deftest]]
             [next.jdbc :as jdbc]
             [xtdb.api :as xt]
             [xtdb.basis :as basis]
@@ -14,7 +14,7 @@
             [xtdb.node.impl] ;;TODO probably move internal methods to main node interface
             [xtdb.object-store :as os]
             [xtdb.protocols :as xtp]
-            [xtdb.serde :as serde] 
+            [xtdb.serde :as serde]
             [xtdb.test-util :as tu]
             [xtdb.time :as time]
             [xtdb.util :as util])
@@ -1199,3 +1199,8 @@ VALUES(1, OBJECT (foo: OBJECT(bibble: true), bar: OBJECT(baz: 1001)))"]])
                       (xt/execute-tx node [[:put-docs :docs {:xt/id 1}]])))
 
     (t/is (= [{:v 1}] (xt/q node "SELECT 1 AS v")))))
+
+(t/deftest test-merge-fields-npe-4721
+  (xt/execute-tx tu/*node* [[:put-docs :docs {:h {:b #{}} :xt/id 1}]])
+  (xt/execute-tx tu/*node* [[:put-docs :docs {:h {}, :xt/id 1} {:xt/id 1}]])
+  (t/is (= [{:xt/id 1}] (xt/q tu/*node* "SELECT * FROM docs"))))

--- a/src/test/clojure/xtdb/types_test.clj
+++ b/src/test/clojure/xtdb/types_test.clj
@@ -370,3 +370,12 @@
     (t/is (= {:vs vs
               :vec-types [RegClassVector RegProcVector]}
              (test-round-trip vs)))))
+
+(t/deftest test-npe-on-empty-list-children-4721
+  (t/testing "merge fields with empty list children shouldn't throw NPE"
+    (let [set-field (types/->field "a" #xt.arrow/type :set true)
+          list-field (types/->field "b" #xt.arrow/type :list true)]
+      (t/is (= (types/->field "a" #xt.arrow/type :set true (types/->field "$data$" #xt.arrow/type :null true)) 
+               (types/merge-fields nil set-field)))
+      (t/is (= (types/->field "b" #xt.arrow/type :list true (types/->field "$data$" #xt.arrow/type :null true)) 
+               (types/merge-fields nil list-field))))))

